### ISSUE TITLE
Serial Assignment Error When using Arduino Leonardo and IDE 1.04

### ISSUE THREAD

**Issue by futureneer** [ros-drivers/rosserial#49]
Hello, I encountered the following error using the 1.04 IDE and a Leonardo when compiling my code against the ros_lib headers:

```
/home/kel/sketchbook/libraries/ros_lib/ArduinoHardware.h:53: error: cannot convert ‘Serial_*’ to ‘HardwareSerial*’ in assignment
```

This was fixed by changing Line 53 of ArduinoHardware.h from 

```
iostream = &Serial;
```

to

```
iostream = &Serial1;
```

as per these instructions:
http://arduino.cc/forum/index.php?topic=111354.0

I am not sure if this is specific to the leonardo, or the 1.04 IDE.  It may have to do with the fact that newer arduinos like the Leonardo have a virtual serial port (more info at: http://arduino.cc/en/Guide/ArduinoLeonardo#toc11).  Just wanted you to know that this is an issue, I am not sure right now how it could be resolved.  Perhaps something like in the above link where the type of arduino is detected and the iostream is assigned a different serial value for each.  This seems kind of messy, but it could work.

I really appreciate your work, it is wonderful to be able to use Arduino on ROS.

Cheers and thanks!
Kel Guerin

**Comment by superxingzheng** [comment]
I solved the same problem by changing the 47th line of ArduinoHardware.h into 
ArduinoHardware(Serial_\* io , long baud= 57600){
and the 80th line into 
Serial_\* iostream;
I think this is the right way to solve this bug. changing into &Serial1 will cause future communication failure.
My method is tested with ROS Fuerte under Ubuntu 12.04 32bit and Arduino Leonardo.

**Comment by mikeferguson** [comment]
I presume there is some sort of #define that tells you that you are compiling for a Leonardo -- if anyone wants to find out what that is and add a #ifdef #else #endif block, and submit a pull request with a patch for this, it'd be much appreciated and certainly merged.

**Comment by PaulBouchier** [comment]
Kel, Zheng,

Thank you for reporting this issue. Can you help me understand your system configuration.

Kel - it seems like you're using a separate USB serial cable from the Leonardo USB cable - and wiring it to ports 0 & 1 on the Arduno. I think that because Serial1 is documented to be the "traditional" serial port, not the virtual serial port on Leonardo.  Is that correct? If so, why do you want to use a separate cable rather than the virtual serial port? Is there a problem with the virtual serial port on Leonardo? (Apologies if I misunderstood - I'm just going off Leonardo documentation).

Zheng, are you using the virtual serial port with your change?

Thanks

Paul

**Comment by SreejitS** [comment]
method suggested by superxingzheng works for me.
Tested under ROS Groovy with Ubuntu 12.04 32bit and Arduino Leonardo.

**Comment by superxingzheng** [comment]
@PaulBouchier Sorry about this late reply! Yes I was using the USBSerial port which Arduino uses to download user codes with its IDE. The previous problem I think is caused by choosing a wrong serial port when switching between different models of Aruduino.

**Comment by SreejitS** [comment]
I think i figured out how to change the baud rate @ which ROS was listening.Changing line no.  49 and 50 in "(workspace)/install/lib/serial_node.py" to the required baud rate and port fixed the issue.
Arduino was publishing @ 115200 and hence making ROS's baud rate to 115200 works out well

**Comment by PaulBouchier** [comment]
Fix released in 0.5.3. New debians should come out in a few days.
